### ISSUE TITLE
Webkit on Raspberry Pi OS

### ISSUE THREAD

**Issue by Resonanz** [beeware/toga#1693]
Toga's Webkit browser component on a Raspberry Pi:

> RuntimeError: Unable to import WebKit2. Ensure that the operating system GTK WebKit bindings (e.g., gir1.2-webkit2-4.0 on Debian/Ubuntu) have been installed.

Fix is:

> sudo apt install gir1.2-webkit2-4.0

**Comment by freakboy3742** [comment]
I'm unclear what you're reporting as a bug and fix.

The tutorial says you need to install `gir1.2-webkit2-4.0`.

The error message tells you you need to install `gir1.2-webkit2-4.0` on a Debian derived OS (as Raspbian is).

And you're reporting... that you need to install `gir1.2-webkit2-4.0`? What am I missing here?

**Comment by Resonanz** [comment]
Sorry, I didn't see the README. 

My expectation was that this would have been installed when I installed Toga.

Forgive my ignorance, but...

...could you explain why the install script does not install it? Is it a deliberate decision, in which case why? 

My guess was that because we are at 0.3.xxx, then it is just something else that needs to be fixed, in which case doing a bug report makes sense to me.

I appreciate all your help and advice so far.

**Comment by freakboy3742** [comment]
Toga is a Python package; Python's installation tooling can only help with installing Python packages.

However, Toga is interacting with the operating system to provide GUI tools. Those tools are provided by the operating system, not by Python.

If you're on macOS or Windows 11, you don't need to install anything extra, because we can make assumptions about what the operating system provides. If you're on Windows 10 you need to install a newer version of Edge to make the web view work. However, we can't assume *anything* is present on Linux. You might not even have a graphical environment.

Making matters worse, there's absolutely no consistency between Linux distributions on how to manage system packages. On Raspbian, you use `apt-get` to insall `gir-1.2-webkit-4.0`. It's also called apt-get on Debian and Ubuntu. But if you're on Fedora, the tool is called `dnf`; if you're on Arch, it's called `pacman`; other Linux distros have other requirements. In addition, those tools need to be run as root, because you're installing system-level libraries.  

In short, there are limits to what Python can do solve this problem. It's an operating system issue, and we (unfortunately) have very limited tools at our disposal to address this. 

Closing this ticket on the basis there's not much we can do to address this, beyond the documentation that is already in place, and the error messages that are already in place telling you what needs to be done.

**Comment by freakboy3742** [comment]
You keep saying "the install script" as if... that was something that exists. 

It doesn't.

pip installs, by preference, from wheels; and wheels, by definition, don't have anything executed at time of installation. This is a security *feature* - because otherwise, it's trivial for malicious code to be executed.

setup.py is a configuration script that can be used to install, if you're installing from source. However, (a) if you're installing Toga, you won't be installing from sources; and (b) PEP517 is an indication that the broader Python community is trying very hard to move away from using setup.py *at all*.

**Comment by Resonanz** [comment]
Thanks. In case others are reading this.......

Wheel (.whl) files are basically zip files with a specific naming convention. 

Running pip unzips the wheel and copies the files into specific locations (which are kinda encoded in the folder names, which is necessary for targeting different platforms).

This video is helpful:

https://www.youtube.com/watch?v=4L0Jb3Ku81s
